### ISSUE TITLE
share configs between multiple apps

### ISSUE THREAD

**Issue by jasonkarns** [lorenwest/node-config#296]
Consider a multiple app scenario where each app needs the same configs (think: cron-worker, queue-worker, app-api, and hosted-express-app for example)

Each of those apps would connect to the same services (db, redis, api services, etc) and need identical configs. (Historically, they could have even been the same "app" with multiple launch/entrypoints)

A thought: with node-config's sub-module config feature, I am thinking about the possibility of making a separate module that is *only* the config. In this way, the app configs would simply be shared across all relevant apps as a dependency. But this scenario would be the reverse of sub-modules configs. Instead of the parent app providing config sections for the submodule, which node-config would handle transparently; rather, the parent module would be pulling the configs from the submodule (since the submodule in this case is merely exporting config).

Does this sound crazy? Alternative ideas?

**Comment by markstos** [comment]
That sounds reasonable to me. I presume you would have a private npm repo or some build automation which would bring in the config dependency as a first step.

**Comment by jasonkarns** [comment]
Yes. What I'm curious of is if there is some wrinkle or feature of
node-config's sub-module capability that would help in this regard. Or if I
should just approach it ignoring submodules entirely, and simply view the
'configs-as-submodule' as an external resource and pull them in/merge
manually as part of bootstrapping.

**Comment by markstos** [comment]
I don't see these as being related. You will be making something like a `my-config` package. From the perspective of each app, they depend on `my-config`. Especially since each one of your apps sounds like a different process, each process is only requiring 'my-config` once, which in turn requires `config` once, so it seems like a pattern that's fairly free of complications, aside from the initial work of breaking our your configs into their own module.

**Comment by jasonkarns** [comment]
After some more thought, I've realized that having `my-config` require `config` is a bad idea. `my-config` should be able to be version bumped solely for configuration changes. If it requires `config`, then it would need version bumped for `config` changes as well, which could contain API changes that are then cascaded to the parent apps. That could cause a scenario where a `config` upgrade blocks the ability to upgrade `my-config`, even if there are legit configuration changes that are necessary.

In short, coupling the version bumps between static configuration data changes and `config` version upgrades is a bad coupling. Rather, I'm thinking the parent app should depend on `config` and `my-config` should just expose the configuration data files. Then the parent app can set `NODE_CONFIG_DIR` to `node_modules/my-config` and be done.

Thanks for the feedback!
